src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -430,4 +465,268 @@ UA_PubSubKeyStorage_detachKeyStorage(UA_Server *server, UA_PubSubKeyStorage *key
+    UA_UNLOCK(&server->serviceMutex);
+    addDelayedSksClientCleanupCb(client);
+    return;
+}
+
+static UA_StatusCode
+callGetSecurityKeysMethod(UA_Client *client) {
+
+    sksClientContext *ctx = (sksClientContext *)client->config.clientContext;
+
+    UA_Variant *inputArguments = (UA_Variant *)UA_calloc(3, (sizeof(UA_Variant)));
+    UA_Variant_setScalarCopy(&inputArguments[0], &ctx->ks->securityGroupID,
+                             &UA_TYPES[UA_TYPES_STRING]);
+    UA_Variant_setScalarCopy(&inputArguments[1], &ctx->startingTokenId,
+                             &UA_TYPES[UA_TYPES_UINT32]);
+    UA_Variant_setScalarCopy(&inputArguments[2], &ctx->requestedKeyCount,
+                             &UA_TYPES[UA_TYPES_UINT32]);
+
+    UA_NodeId objectId = UA_NODEID_NUMERIC(0, UA_NS0ID_PUBLISHSUBSCRIBE);
+    UA_NodeId methodId = UA_NODEID_NUMERIC(0, UA_NS0ID_PUBLISHSUBSCRIBE_GETSECURITYKEYS);
+    size_t inputArgumentsSize = 3;
+
+    return UA_Client_call_async(client, objectId, methodId, inputArgumentsSize,
+                                inputArguments, storeFetchedKeys, (void *)ctx, NULL);
+}
+
+static void
+onConnect(UA_Client *client, UA_SecureChannelState channelState,
+          UA_SessionState sessionState, UA_StatusCode connectStatus) {
+    UA_Boolean triggerSKSCleanup = false;
+    if(connectStatus != UA_STATUSCODE_GOOD && channelState == UA_SECURECHANNELSTATE_CLOSED && sessionState == UA_SESSIONSTATE_CLOSED) {
+        UA_LOG_ERROR(&client->config.logger, UA_LOGCATEGORY_CLIENT,
+                     "SKS Client: Failed to connect SKS server with error: %s ",
+                     UA_StatusCode_name(connectStatus));
+        triggerSKSCleanup = true;
+    }
+    if(connectStatus == UA_STATUSCODE_GOOD && sessionState == UA_SESSIONSTATE_ACTIVATED) {
+        connectStatus = callGetSecurityKeysMethod(client);
+        if(connectStatus != UA_STATUSCODE_GOOD) {
+            UA_LOG_ERROR(&client->config.logger, UA_LOGCATEGORY_SERVER,
+                         "SKS Client: Failed to call GetSecurityKeys on SKS server with "
+                         "error: %s ",
+                         UA_StatusCode_name(connectStatus));
+            triggerSKSCleanup = true;
+        }
+    }
+    if(triggerSKSCleanup) {
+        client->noReconnect = true;
+        /* call user callback to notify about the status */
+        sksClientContext *ctx = (sksClientContext *)client->config.clientContext;
+        UA_PubSubKeyStorage *ks = ctx->ks;
+        if(ks->sksConfig.sksPullRequestCallback)
+            ks->sksConfig.sksPullRequestCallback(ctx->server, connectStatus,
+                                            ks->sksConfig.userData);
+        addDelayedSksClientCleanupCb(client);
+    }
+    return;
+}
+
+static void
+setExternalEvenloopOnSkSClient(UA_ClientConfig *cc, UA_EventLoop *externalEventloop) {
+    UA_assert(externalEventloop != NULL);
+
+    /* Stop and delete the client EventLoop */
+    UA_EventLoop *el = cc->eventLoop;
```

Comment:
Stopping the event loop can have unintended side effects.
As it could be configured for multiple applications at once.

Why is this needed? SKS should be able to initially set an already running EventLoop.
Switching the EventLoop mid-flight of the SKS opens a can of worms. I wouldn’t support that.

If the EventLoop has to be stopped, log an error and return.
But don’t stop the EventLoop on your own.
Developers will not be aware of this without having read the code.

Reply:
> Stopping the event loop can have unintended side effects. As it could be configured for multiple applications at once.
> 
> Why is this needed? SKS should be able to initially set an already running EventLoop. Switching the EventLoop mid-flight of the SKS opens a can of worms. I wouldn’t support that.
> 
The eventloop is not switched in mid flight but immediately after creating the sks client instance. This allows sks client to use the server's eventloop and the SKS Client lifecycle is managed in UA_server_run_iterate calls. This is all happening in `getSecurityKeysAndStoreFetchedKeys`

> If the EventLoop has to be stopped, log an error and return. But don’t stop the EventLoop on your own. Developers will not be aware of this without having read the code.

Before stopping the eventloop there is a check for externalEventLoop flag. If it is set then eventloop is not stopped and simply replaced. But if the eventloop is not shared/external then stopping is important, otherwise the eventloop pointer will be lost and cannot be freed.